<h1><center>Laboratorio 9: Los huesos de Hipócrates 🦴</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesor: Matías Rojas y Mauricio Araneda
- Auxiliar: Ignacio Meza D.
- Ayudante: Rodrigo Guerra

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Emilio Roa
- Nombre de alumno 2:


### **Link de repositorio de GitHub:** `https://github.com/ERoa98/MDS7202-Emilio-Roa`

### Indice 

1. [Temas a tratar](#Temas-a-tratar:)
3. [Descripcción del laboratorio](#Descripción-del-laboratorio.)
4. [Desarrollo](#Desarrollo)

## Temas a tratar

- Creación de clasificadores de imagenes a traves de redes Fully connected y CNN.
- Uso de Dataloaders para la carga de datasets.
- Comparación de Fully Connected y red convolucional.

## Reglas:

- Fecha de entrega: 17/11/2022
- **Grupos de 2 personas**
- **Ausentes** deberán realizar la actividad solos. 
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias. 
- Pueden usar cualquier material del curso que estimen conveniente.

### Objetivos principales del laboratorio

- Creación de modelos de clasificación de imágenes utilizando Pytorch.
- Creación de dataloader y aplicar transformaciones sobre el dataset.
- Comprender la diferencia entre una CNN y una Fully Connected.

El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones de `Pytorch`, la cual, está enfocada para proyectos de Deep Learning.


# Importamos librerias utiles 😸

Comenzamos importando librerías utiles para la ejecución del laboratorio:

In [1]:
import os
import time
from PIL import Image

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import plotly.express as px

import torch
import torch.nn as nn
import torch.nn.functional as F

import torch.optim as optim
from torch.optim import lr_scheduler
from torch.autograd import Variable
from torch.utils.data import Dataset, DataLoader

import torchvision
from torchvision import datasets, models
from torchvision import transforms as T

# Identificando los Huesos de Hipócrates🔎

<p align="center">
  <img src="https://i.pinimg.com/originals/d8/58/66/d85866cd1cc3979f94526551addf74b4.gif" width="300">
</p>

Tras el éxito que han tenido proponiendo modelos de machine learning en trabajos anteriores, el famoso medico Hipócrates se ha contactado con ustedes para solicitarles ayuda para automatizar la identificación de radiografías de partes humanas. Para esto, les señala que le gustaría utilizar algoritmos de deep learning producto que Demócrito le señalo que resultan la mejor alternativa para la predicción de imágenes.

En su conversación con el medico usted le comenta que ha tenido algunas clases relacionadas a Deep Learning, por esto, están motivados en abordar el problema utilizando redes Fully Connected y redes convolucionales con Pytorch. Sin embargo, al anunciarle los tipos de redes que conocen, el filósofo les comenta que no había escuchado muy buenos resultados por parte de las CNN, por lo que les pide que comprueben a traves de la métrica de accuracy que tipo de redes es mejor para la tarea de identificación de radiografías. ¿Será cierto lo que dice el filósofo?, Veámoslo en un nuevo capítulo de los Laboratorios de Programación Científica para Ciencia de Datos!

## 1.1 Creación de Lista de Archivos

<p align="center">
  <img src="https://media.tenor.com/BJ-9w-MUVCMAAAAM/tis100-sad.gif" width="300">
</p>


Comience revisando de forma manual cada una de las imágenes que posee la carpeta subida a material docente. Verifique la cantidad de tipos de radiografías que se tienen y la cantidad de imágenes que dispone cada carpeta.

Revisado el contenido de las imágenes, utilice `os.listdir` para crear un `numpy.array` o un `Dataframe` que contenga las imágenes y un label que señale al tipo de radiografía a la que hace referencia la imagen. Para hacer las etiquetas codifique el tipo de imágenes en números que vayan del 0 al total de tipos de radiografías, no utilice strings para codificar las etiquetas.

**Ejemplo de Estructura:**

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-0lax{text-align:left;vertical-align:top}
</style>
<table class="tg">
<thead>
  <tr>
    <th class="tg-0lax"></th>
    <th class="tg-0lax">image_path</th>
    <th class="tg-0lax">label</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-0lax">0</td>
    <td class="tg-0lax">image1</td>
    <td class="tg-0lax">1</td>
  </tr>
  <tr>
    <td class="tg-0lax">1</td>
    <td class="tg-0lax">image2</td>
    <td class="tg-0lax">0</td>
  </tr>
  <tr>
    <td class="tg-0lax">2</td>
    <td class="tg-0lax">image3</td>
    <td class="tg-0lax">2</td>
  </tr>
  <tr>
    <td class="tg-0lax">3</td>
    <td class="tg-0lax">image4</td>
    <td class="tg-0lax">0</td>
  </tr>
  <tr>
    <td class="tg-0lax">4</td>
    <td class="tg-0lax">image5</td>
    <td class="tg-0lax">4</td>
  </tr>
</tbody>
</table>

**Respuesta:**

Al revisar de forma manual las carpetas con las imágenes se puede comprobar que se tiene un total de 58954 imágenes, las cuales se dividen en 6 clases diferentes donde cada clase contiene:

- AbdomenCT: 10.000
- BreastMRI: 8.954
- ChestCT: 10.000
- CXR: 10.000
- Hand: 10.000  
- HeadCT: 10.000

A continuación se almacenan los archivos de las carpetas en un dataframe almacenando el path de cada imagen, un arreglo con los valores de la imagen y las labels respectivas.

In [2]:
# Código Aquí

images = []
categories = []
img_paths = []

#path = "/media/roa/Data/Primavera 2022/Laboratorio de Programación Cientifica/2022111011141005A0888D541A93__Medical-MNIST/Medical-MNIST"
path = "D:\Primavera 2022\Laboratorio de Programación Cientifica\Medical-MNIST\Medical-MNIST"
for i, category in enumerate(os.listdir(path)):
    print(category)
    for image in os.listdir(f"D:\Primavera 2022\Laboratorio de Programación Cientifica\Medical-MNIST\Medical-MNIST/{category}"):
        path = f"D:\Primavera 2022\Laboratorio de Programación Cientifica\Medical-MNIST\Medical-MNIST/{category}/{image}"
        img = plt.imread(path)
        images.append(img)
        categories.append(category)
        img_paths.append(path)
        

AbdomenCT
BreastMRI
ChestCT
CXR
Hand
HeadCT


In [3]:
df = pd.DataFrame()
df['img_paths'] = img_paths
df['img'] = images
df['categories'] = categories

dic = {'AbdomenCT': 0, 'BreastMRI': 1, 'ChestCT': 2, 'CXR': 3, 'Hand': 4, 'HeadCT': 5}

df['labels'] = df['categories'].map(dic)

df

img_paths  \
0      D:\Primavera 2022\Laboratorio de Programación ...   
1      D:\Primavera 2022\Laboratorio de Programación ...   
2      D:\Primavera 2022\Laboratorio de Programación ...   
3      D:\Primavera 2022\Laboratorio de Programación ...   
4      D:\Primavera 2022\Laboratorio de Programación ...   
...                                                  ...   
58949  D:\Primavera 2022\Laboratorio de Programación ...   
58950  D:\Primavera 2022\Laboratorio de Programación ...   
58951  D:\Primavera 2022\Laboratorio de Programación ...   
58952  D:\Primavera 2022\Laboratorio de Programación ...   
58953  D:\Primavera 2022\Laboratorio de Programación ...   

                                                     img categories  labels  
0      [[101, 101, 101, 101, 101, 101, 101, 101, 101,...  AbdomenCT       0  
1      [[101, 101, 101, 101, 101, 101, 101, 101, 101,...  AbdomenCT       0  
2      [[101, 101, 101, 101, 101, 101, 101, 101, 101,...  AbdomenCT       0  
3      [[101, 101, 101, 101, 101, 101, 101, 101, 101,...  AbdomenCT       0  
4      [[101, 101, 101, 101, 101, 101, 101, 101, 101,...  AbdomenCT       0  
...                                                  ...        ...     ...  
58949  [[25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, ...     HeadCT       5  
58950  [[25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, ...     HeadCT       5  
58951  [[25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, ...     HeadCT       5  
58952  [[25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, ...     HeadCT       5  
58953  [[25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, ...     HeadCT       5  

[58954 rows x 4 columns]

In [4]:
s = df.iloc[0]
img_path = s.loc['img_paths']

img_path

'D:\\Primavera 2022\\Laboratorio de Programación Cientifica\\Medical-MNIST\\Medical-MNIST/AbdomenCT/000000.jpeg'

## 1.2 Creación de Dataset

Tomando en cuenta la estructura de datos desarrollada en el punto 1.1, construya la clase `MedicalDataset()` que cumpla los siguientes puntos:

- [ ] Poseer un `__init__` en el que se almacene `estructura` creada en 1.1, la `raiz` de la carpeta y una función que permita transformar el dataset (de esto no se preocupe mucho, ya que solamente debe almacenar una función en el atributo).
- [ ] La clase debe ser capaz de entregar la cantidad de elementos a traves de `__len__`.
- [ ] Debe poseer el método `__getitem__` que retorne una tupla con la imagen y su correspondiente etiqueta.

In [5]:
# Código Aquí
class MedicalDataset(Dataset):
    def __init__(self, df, raiz, transform):
        self.df = df
        self.raiz = raiz
        self.transform = transform
    
    def __getitem__(self, idx):
        s = self.df.iloc[idx]
        img_path = s.loc['img_paths']
        
        # Un poco de ayuda para cargar la imagen
        image = Image.open(img_path).convert('RGB')
        label = s.loc['labels']
        
        # Auida para realizar la transformación
        if self.transform:
            image = self.transform(image)
        
        return image, label
    
    def __len__(self):
        return len(self.df)

## 1.3 Prueba del MedicalDataset

Con la clase construida en el punto 1.2, verifique su funcionamiento cargando el dataset y realizando las transformaciones que entrega la función `transform_image`. Compruebe a través de un ejemplo las transformaciones aplicadas en la imagen, comentando la función que cumple `MedicalDataset` y si es posible observar todas las transformaciones aplicadas con la función `transform_image`.

- [ ] Probar la clase MedicalDataset y aplicando una transformación de "train"
- [ ] Plotear un ejemplo del MedicalDataset.

**Función para transformar las imagenes:**

In [6]:
def transform_image(stage = None):
    
    if stage == "train":
        Tr_img = T.Compose([T.Resize(size = (128,128)),
                T.RandomRotation(degrees = (-20,+20)),
                T.ToTensor()])
        
    elif stage == "test" or stage == "val":
        Tr_img = T.Compose([T.Resize(size = (128,128)), T.ToTensor()]) 

    return Tr_img

**Código para obtener un ejemplo:**

In [7]:
# Prueba del dataset
ruta = "D:\Primavera 2022\Laboratorio de Programación Cientifica\Medical-MNIST\Medical-MNIST"
dataset = MedicalDataset(df, ruta, transform_image("train"))

In [8]:
data_test= iter(dataset)
example_image = next(data_test)

# Utilice plotly para plotear un ejemplo
img = example_image[0].numpy()[0]
fig = px.imshow(img)
fig.show()

> Comente que realiza la clase construida y las transformaciones aplicadas.

La clase contruida MedicalDataset, al heredar de la clase Dataset de pytorch nos permite construir nuestro propio dataset y heredar todos los métodos disponibles en pytorch para trabajar con datasets, como por ejemplo construir dataloaders, manejar los datos de forma más eficiente o aplicar transformaciones de forma directa. En cuanto a las transformaciones aplicadas, estas se pueden observar al visualizar el ejemplo.

El tamaño original de las imágenes del dataser es 64x64, sin embargo al aplicar el resize se puede verificar que las nuevas dimensiones son de 128x128. Además, se aplicó una rotación a las imágenes de train, lo que de igual forma se puede observar claramente en el ejemplo.

## 1.4 Creación de Clasificadores

<p align="center">
  <img src="https://149695847.v2.pressablecdn.com/wp-content/uploads/2018/01/conv-full-layer.gif" width="300">
</p>

A continuación, deben construir tres clasificadores con los que deberán verificar cuál de las arquitecturas posee un mejor desempeño para la tarea de clasificación de imágenes. Para la construcción considere los siguientes puntos:

- [ ] Señale cual es el objetivo del `forward` en este tipo de redes, sea breve para su explicación.
- [ ] Construir una red Fully Connected para solucionar el problema de clasificación. Para esta parte se le aconseja que rellene el esqueleto dispuesto más abajo y que lleva el nombre de `FCClassifier`, en el deberá rellenar con la dimensión de las capas ocultas y verificar cual será el tamaño de la entrada.
- [ ] Construya una red convolucional **simple** (no más de una capa convolucional) para la tarea de clasificación de imágenes, para esto basen su código en la clase del día `09-11-2022`.
- [ ] Crear una red convolucional más compleja. Para esta parte tienen completa libertad en la construcción de su red, lo único que debe cumplir es que sea convolucional.

**Esqueletos Propuestos:**

**Respuesta:**

El objetivo del `forward` en este tipo de redes neuronales es calcular el output de la red con las valores de parámetros actuales haciendo todas las multiplicaciones matriciales correspondientes a la red, para así con este output poder obtener un error al comparar el output obtenido con el output deseado.

In [9]:
class FCClassifier(nn.Module):
    def __init__(self, in_channels, num_classes):
        super(FCClassifier, self).__init__()
        self.flatten = nn.Flatten()
        self.lin1 = nn.Linear(in_channels, 16)
        self.lin2 = nn.Linear(16, num_classes)
        
    def forward(self, x):
        batch_size = x.shape[0]
        x = x.view(batch_size, -1)
        h_1 = F.relu(self.lin1(x))
        y_pred = self.lin2(h_1)
    
        return y_pred

In [10]:
class CNN_Net(nn.Module):
    
    def __init__(self):
        super(CNN_Net, self).__init__()
        
        self.layer1 = nn.Sequential(
            nn.Conv2d(in_channels=3, out_channels=8, kernel_size=3, padding=1), #128*128*3 -> 128x128x8
            nn.BatchNorm2d(8),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2) #64x64x8
        )
        
        self.fc1 = nn.Linear(in_features=64*64*8, out_features=100)
        self.drop = nn.Dropout2d(0.25)
        self.fc2 = nn.Linear(in_features=100, out_features=50)
        self.fc3 = nn.Linear(in_features=50, out_features=6)
        
    def forward(self, x):
        out = self.layer1(x)
        out = out.view(out.size(0), -1)
        out = self.fc1(out)
        out = self.drop(out)
        out = self.fc2(out)
        out = self.fc3(out)
        
        return out

class CNN_Net2(nn.Module):
    
    def __init__(self):
        super(CNN_Net2, self).__init__()
        
        self.layer1 = nn.Sequential(
            nn.Conv2d(in_channels=3, out_channels=8, kernel_size=3, padding=1), #128*128*3 -> 128x128x8
            nn.BatchNorm2d(8),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2) #64x64x8
        )
        
        self.layer2 = nn.Sequential(
            nn.Conv2d(in_channels=8, out_channels=16, kernel_size=3), #62x62x16
            nn.BatchNorm2d(16),
            nn.ReLU(),
            nn.MaxPool2d(2) #31x31x16
        )
        
        self.fc1 = nn.Linear(in_features=16*31*31, out_features=100)
        self.drop = nn.Dropout2d(0.25)
        self.fc2 = nn.Linear(in_features=100, out_features=50)
        self.fc3 = nn.Linear(in_features=50, out_features=6)
        
    def forward(self, x):
        out = self.layer1(x)
        out = self.layer2(out)
        out = out.view(out.size(0), -1)
        out = self.fc1(out)
        out = self.drop(out)
        out = self.fc2(out)
        out = self.fc3(out)
        
        return out


class CNNClassifier1(nn.Module):
    def __init__(self, in_channels, num_classes):
        super(CNNClassifier1, self).__init__()
        self.conv1 = nn.Conv2d(in_channels , 4, kernel_size=3, padding=1)
        self.fc1 = nn.Linear(64*64*4, 32)
        self.fc2 = nn.Linear(32, 16)
        self.fc3 = nn.Linear(16,num_classes)
    def forward(self, x):
        # x es 128 x 128
        x = F.relu(F.max_pool2d(self.conv1(x), 2)) # conv: 128x128x4 -> maxpool: 64x64x4
        x = x.view(-1, 64*64*4)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

    
class CNNClassifier2(nn.Module):
    def __init__(self, in_channels, num_classes):
        super(CNNClassifier2, self).__init__()
        self.conv1 = nn.Conv2d(in_channels, 16, kernel_size=3, padding=1) 
        self.conv2 = nn.Conv2d(16, 32, kernel_size=3, padding=1)
        self.conv3 = nn.Conv2d(32, 64, kernel_size=3)
        self.fc1 = nn.Linear(30*30*64, 64)
        self.fc2 = nn.Linear(64, num_classes)
    def forward(self, x):
        # x es 128 x 128
        x = F.relu(F.max_pool2d(self.conv1(x), 2))# conv: 128x128x16 -> maxpool: 64x64x16
        x = F.relu(F.max_pool2d(self.conv2(x), 2))# conv: 64x64x32 -> maxpool: 32x32x32
        x = F.relu(self.conv3(x))# conv: 30x30x64
        x = x.view(-1, 30*30*64)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        return x
        #return F.log_softmax(x)

## 1.5 Separando Datos para el Entrenamiento

<p align="center">
  <img src="https://c.tenor.com/Esn7Jif-ZLQAAAAC/separate-square.gif" width="200">
</p>

Utilizando un Holdout a su gusto, separe los datos en un conjunto de entrenamiento y de testing. Aplique las transformaciones correspondientes usando `transform_image` para cada conjunto de datos y utilice `torch.utils.data.DataLoader` para crear un objeto iterable del dataset.

- [ ] Definir el Holdout a utilizar.
- [ ] Separar los datos en un conjunto de entrenamiento y prueba.
- [ ] Aplicar las transformaciones correspondientes en cada uno de los dataset.
- [ ] Utilizar `DataLoader` de pytorch sobre los dataset.

In [11]:
# Separar Datos de Entrenamiento
from sklearn.model_selection import train_test_split

train, test = train_test_split(df, test_size=0.3)

train_dataset = MedicalDataset(train, ruta, transform_image("train"))
test_dataset = MedicalDataset(test, ruta, transform_image("test"))

In [12]:
# Uso de torch.utils.data.DataLoader
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader= torch.utils.data.DataLoader(test_dataset, batch_size=64, shuffle=True)

## 1.6 Creación de Funciones de Entrenamiento y Evaluación

<p align="center">
  <img src="https://www.researchgate.net/publication/319535615/figure/fig3/AS:536187598065664@1504848493070/A-typical-convolutional-neural-network-CNN-Architecture-for-Medical-Image-Classification.png" width="500">
</p>


Ya construido todas las funciones y clases necesarias llego el momento más importante... probar la red. Para esta sección, ustedes deberán ser capaces de definir los hiperparámetros de la red, definir las funciones de perdida a utilizar, señalar el optimizador a usar y finalmente crear sus funciones para el entrenamiento y prueba. Para realizar esta parte más estructurada, seguir los siguientes puntos de forma secuencial:

- [ ] Especifique los Hiperparámetros de las 3 redes. Para esta parte sea claro de su elección y señale el porqué de sus elecciones (o sea justifique el setting de sus hiperparámetros).
- [ ] Defina los modelos a utilizar, el optimizador que utilizará para el modelo y señale la función de perdida que utilizará.
- [ ] Explique de forma breve la función que cumplen los pasos `Backward` y `Descenso del gradiente` en una red neuronal.
- [ ] Cree una función llamado `train` que entrene a los clasificadores. Para esto, recuerde que estos modelos suelen utilizar un número de épocas, por lo que deberá generar un proceso iterativo de entrenamiento. Es importante que su función imprima las `loss` obtenidas por el modelo en cada época (si gusta puede almacenar estas losses en una lista para luego graficarlas y comparar).
- [ ] Diseñe una función para evaluar el desempeño de las redes. Para evaluar las redes utilice solamente la métrica accuracy (para esto se le recomienda comparar la predicción con el ground truth)

In [13]:
# Especificar hyperparámetros de las redes
in_channels_fc = 128*128*3 # los canales de entrada para la red feedforward se obtienen al sumar todos los parametros que tienen las imágenes, en este caso alto por ancho por el número de canales
in_channels_cnn = 3 # para las redes cnn los canales de entrada hacen referencia solo a los canales de colores
num_classes = 6 
lr = 0.001 # se eligió un lr = 0.001 ya que debe ser un valor pequeño para poder converger de buena manera durante la actualización de los pesos, y 0.001 es un valor bastante usado en la literatura
batch_size = 64 # un batchsize de 64 ya que el dataset es bastante grande
n_epochs = 5 # solo 5 épocas porque se demora mucho en entrenar y no tengo gpu :(

device = ('cuda' if torch.cuda.is_available() else 'cpu')


In [18]:
# Red 1
model_fc = FCClassifier(in_channels_fc, num_classes)
model_cnn1 = CNNClassifier1(in_channels_cnn, num_classes)
model_cnn2 = CNNClassifier2(in_channels_cnn, num_classes)
model_3 = CNN_Net()
model_4 = CNN_Net2()

# Loss and optimizer
criterion = nn.CrossEntropyLoss() # se define como función de loss la crossentropy ya que tiene buenos buenos desempeños en este tipo de problemas y es bastante usado en problemas similares
optimizer = optim.SGD(model_3.parameters(), lr=lr, momentum=0.9) # se utilizó un gradiente descendiente estocástico para asegurar una buena convergencia de los modelos




**Respuesta:**

En la etapa de backward se realiza el proceso de backpropagation, en donde en base al error encontrado entre el output obtenido y el deseado se propaga el error hacia atrás calculando las derivadas y cálculos respectivos para de esta forma saber que parámetros o que partes de la red están más relacionados con el error.

En base a esto posteriormente se puede aplicar el el algoritmo de gradiente descendiente en donde se actualizan los pesos de la red para reducir el error en base a los valores de los gradientes previamente calculados.

In [15]:
# train
def train(model, train_loader, criterion, optimizer, n_epochs):
    since = time.time()

    train_loss = []
    train_accu = []
    
    for e in range(n_epochs):
        model.train()
        running_loss = 0.0
        running_corrects = 0.0
        for i, (data, label) in enumerate(train_loader):
            data = data
            targets = label
            
            #Forward
            optimizer.zero_grad()
            outputs = model(data)
            #print('output',outputs)
            #print('label',label)
            _, preds = torch.max(outputs, 1)
            
            #Backward
            loss = criterion(outputs, label)
            #loss = F.nll_loss(outputs, label)
            loss.backward()
            
            # Descenso del gradiente
            optimizer.step()
            running_loss += loss.item() * data.size(0)
            running_corrects += torch.sum(preds == label.data)

        epoch_loss = running_loss /len(train_dataset)
        epoch_acc = running_corrects.double() / len(train_dataset)

        train_loss.append(epoch_loss)
        train_accu.append(epoch_acc)

        print('Train Loss: {:.4f} Acc: {:.4f}'.format(epoch_loss, epoch_acc))

    time_elapsed = time.time() - since
    print('Training complete in {:.0f}m {:.0f}s'.format(time_elapsed//60, time_elapsed % 60))

    return model, train_loss, train_accu

def calculate_accuracy(y_pred, y):
  top_pred = y_pred.argmax(1, keepdim=True)
  correct = top_pred.eq(y.view_as(top_pred)).sum()
  acc = correct.float()/y.shape[0]
  return acc

# Evaluate
def evaluate(loader, model, criterion, device):
    epoch_loss = 0
    epoch_acc = 0

    model.eval()
    with torch.no_grad():
        for (x, y) in loader:
            x = x.to(device)
            y = y.long().to(device)

            y_pred = model(x)

            loss = criterion(y_pred, y)

            acc = calculate_accuracy(y_pred, y)

            epoch_loss += loss.item()
            epoch_acc += acc.item()
    return epoch_loss / len(loader), epoch_acc / len(loader)


## 1.7 Comparación de Resultados

<p align="center">
  <img src="https://media2.giphy.com/media/icJA0VF7ntoEL18Jez/giphy.gif"  width="200">
</p>

Construidas las funciones de entrenamiento y evaluación, entrene a las redes que construyo y compare los resultados obtenidos con todas las redes señalando cual posee mejor rendimiento. Comente una diferencia entre las redes Fully Connected y CNN podría generar un mejor desempeño en una u otra en la tarea de clasificación de imágenes.

- [ ] Entrenar las redes.
- [ ] Evaluar las redes.
- [ ] Comentar los resultados obtenidos y visualizar si existe una diferencia significativa en el rendimiento debido a la naturaleza de la red.

In [191]:
# Código aquí
model, train_loss, train_accu = train(model_fc, train_loader=train_loader, criterion= criterion, optimizer = optimizer, n_epochs=5)

Train Loss: 0.3962 Acc: 0.8839
Train Loss: 0.1552 Acc: 0.9648
Train Loss: 0.1237 Acc: 0.9707
Train Loss: 0.1056 Acc: 0.9743
Train Loss: 0.0953 Acc: 0.9754
Training complete in 4m 60s


In [192]:
evaluate(test_loader, model, criterion=criterion, device=device)

(0.07970751175784678, 0.9785085740072202)

In [277]:
model, train_loss, train_accu = train(model_3, train_loader=train_loader, criterion= criterion, optimizer = optimizer, n_epochs=5)
evaluate(test_loader, model, criterion=criterion, device=device)

Train Loss: 0.0619 Acc: 0.9850
Train Loss: 0.0361 Acc: 0.9904
Train Loss: 0.0238 Acc: 0.9942
Train Loss: 0.0187 Acc: 0.9952
Train Loss: 0.0139 Acc: 0.9966
Training complete in 17m 7s


(0.008416981441618131, 0.9980821299638989)

In [19]:
optimizer = optim.SGD(model_4.parameters(), lr=lr, momentum=0.9)
model, train_loss, train_accu = train(model_4, train_loader=train_loader, criterion= criterion, optimizer = optimizer, n_epochs=5)
evaluate(test_loader, model, criterion=criterion, device=device)

c:\Users\emili\AppData\Local\Programs\Python\Python38\lib\site-packages\torch\nn\functional.py:1331: UserWarning:

dropout2d: Received a 2-D input to dropout2d, which is deprecated and will result in an error in a future release. To retain the behavior and silence this warning, please use dropout instead. Note that dropout2d exists to provide channel-wise dropout on inputs with 2 spatial dimensions, a channel dimension, and an optional batch dimension (i.e. 3D or 4D inputs).



Train Loss: 0.0827 Acc: 0.9782
Train Loss: 0.0161 Acc: 0.9965
Train Loss: 0.0100 Acc: 0.9979
Train Loss: 0.0071 Acc: 0.9982
Train Loss: 0.0062 Acc: 0.9984
Training complete in 35m 21s


(0.004908428779289014, 0.9989282490974729)

> Comente los resultados

Una vez ya entrenadas y evaluadas cada una de las 3 redes, es posible notar que las 3 obtienen resultados super buenos siendo la que tiene peor desempeño la red fullyconected con un accuracy de test de 97%, en segundo lugar la red convolucional simple con un 99.80% y la red con mejor desempeño fue la segundo red convolucional con un 99,89% de accuracy de test.

Estos resultados nos sugiere que claramente el problema era simple de resolver, y si observamos ejemplos de imágenes de cada una de las 6 clases es posible notar que cada clase es muy diferenciable con respecto a las demás. Sin embargo, a pesar de que las 3 redes obtuvieron exelentes accuracis es posible notar que existe una diferencia de rendimiento de 2% entre la red fullyconnected y las redes convolucionales. Este mejor desempeño de las CNN es gracias a su arquitectura ya que al tener capas convolucionales pueden procesar imágenes de mejor forma capturando la información espacial a través de filtros, lo cual es de vital importancia para procesar imágenes. En cambio la red fullt connected no puede capturar esta información y solo lee las imágenes como un gran vector de números sin poder relacionar los valores con la ubicación dentro de la imagen.

# Conclusión

Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media.tenor.com/vKSR-ZakVMIAAAAC/pochitadancing-pochita.gif">
</p>